src/scalarstats.jl

```diff
@@ -260,37 +260,80 @@ variation(x) = ((m, s) = mean_and_std(x); s/m)
 realXcY(x::Real, y::Real) = x*y
 realXcY(x::Complex, y::Complex) = real(x)*real(y) + imag(x)*imag(y)
 
+
 """
-    sem(x)
+    sem(x[, weights::AbstractWeights]; mean=nothing)
+
+Return the standard error of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
+
+# References
 
-Return the standard error of the mean of collection `x`,
-i.e. `sqrt(var(x, corrected=true) / length(x))`.
+Carl-Erik Sarndal, Bengt Swensson, Jan Wretman (1992). Model Assisted Survey Sampling (pp. 51-53). ISBN 9780387975283. 
 """
-function sem(x)
-    y = iterate(x)
-    if y === nothing
-        T = eltype(x)
-        # Return the NaN of the type that we would get, had this collection
-        # contained any elements (this is consistent with std)
-        return oftype(sqrt((abs2(zero(T)) + abs2(zero(T)))/2), NaN)
+function sem(x; mean=nothing)
+    n = 0
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        variance = var(x; mean=mean, corrected=corrected)
+        mean = Statistics.mean(x)
+    elseif mean ≢ nothing
+        sse = real(zero(mean))
+        @simd for element in x
+            n += 1
+            sse += abs2(element - mean)
+        end
+        variance = sse / (n - corrected)
+    else
+        # Use Welford algorithm as seen in (among other places)
+        # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
+        mean = zero(eltype(x))
+        sse = real(mean)
+        for element in x
+            n += 1
+            new_mean = mean + (element - mean) / n
+            sse += realXcY(element - mean, element - new_mean)
+            mean = new_mean
+        end
+        variance = sse / (n - corrected)
     end
-    count = 1
-    value, state = y
-    y = iterate(x, state)
-    # Use Welford algorithm as seen in (among other places)
-    # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
-    M = value / 1
-    S = real(zero(M))
-    while y !== nothing
-        value, state = y
-        y = iterate(x, state)
-        count += 1
-        new_M = M + (value - M) / count
-        S = S + realXcY(value - M, value - new_M)
-        M = new_M
+    return sqrt(variance / n)
+end
+
+function sem(x::RealArray; mean=nothing)
+    var = var(x; mean=mean, corrected=true)
+    return sqrt(var / length(x))
+end
```

Comment:
This will yield a `UndefVarError`.
```suggestion
sem(x::RealArray; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))
```